File: javascript/17_lista/lista.js
const lista = (function() {
	var cabeca = {};
	var topo = 0;
	
	function empilhar(valor) {
		var novo = {dado: valor, proximo: null};
		
		if (topo == 0) {
			cabeca = novo;
		} else {
			var corrente = cabeca;
			
			while (corrente.proximo != null) {
				corrente = corrente.proximo;
			}
			
			corrente.proximo = novo;
		}
		
		topo++;
	}
	
	function inserir(valor, posicao) {
		if (posicao >=0 && posicao <= topo) {
			var novo = {dado: valor, proximo: null};
			
			if (posicao == 0) {
				novo.proximo = cabeca;
				cabeca = novo;
			} else {
				var antecessor = obter(posicao - 1);
				var sucessor = antecessor.proximo;
				
				antecessor.proximo = novo;
				novo.proximo = sucessor;				
			}
			
			topo++;
			
			return true;
		}
		
		return false;
	}
	
	function remover(posicao) {
		if (posicao == 0) {
			cabeca = cabeca.proximo;
		} else {
			var antecessor = obter(posicao - 1);
			var corrente = antecessor.proximo;
			var sucessor = corrente.proximo;
			
			antecessor.proximo = sucessor;
		}
		
		topo--;
	}
	
	function obter(posicao) {
		var elemento = cabeca;
		
		for (var i = 0; i < posicao; i++) {
			elemento = elemento.proximo;
		}
		
		return elemento;
	}
	
	function imprimir() {
		if (topo > 0) {
			var corrente = cabeca;
			
			document.write(corrente.dado + " ");
			
			while (corrente.proximo != null) {
				corrente = corrente.proximo;
				document.write(corrente.dado + " ");
			}
		}
	}
	
	function tamanho() {
		return topo;
	}
	
	function inserirEmOrdem(valor) {
		var corrente = cabeca;
		
		if (topo == 0) {
			inserir(valor, 0);
			return;
		}
		
		var i = 0;
		
		for (; i < topo; i++) {
			if (valor < corrente.dado) {
				break;
			}
			
			corrente = corrente.proximo;
		}
		
		inserir(valor, i);
	}
	
	return {
		empilhar: empilhar,
		inserir: inserir,
		remover: remover,
		obter: obter,
		imprimir: imprimir,
		tamanho: tamanho,
		inserirEmOrdem: inserirEmOrdem
	}
}()); 

File: python/_modulos/lista.py
from dataclasses import dataclass
@dataclass
class no:
    dado: str
    proximo: dataclass

class Lista:    
    def __init__(self):
        self.cabeca = None
        self.topo = 0

    def empilhar(self, valor):
        novo = no(valor, None)

        if self.topo == 0:
            self.cabeca = novo
        else:
            corrente = self.cabeca

            while corrente.proximo != None:
                corrente = corrente.proximo

            corrente.proximo = novo

        self.topo += 1

    def obter(self, posicao):
        elemento = self.cabeca

        for i in range(posicao):
            elemento = elemento.proximo

        return elemento
    
    def inserir(self, valor, posicao):
        if (posicao >= 0) and (posicao <= self.topo):
            novo = no(valor, None)

            if posicao == 0:
                novo.proximo = self.cabeca
                self.cabeca = novo
            else:
                antecessor = self.obter(posicao - 1)
                sucessor = antecessor.proximo

                antecessor.proximo = novo
                novo.proximo = sucessor

            self.topo += 1
            return True
        
        return False
    
    def remover(self, posicao):
        if posicao == 0:
            self.cabeca = self.cabeca.proximo
        else:
            antecessor = self.obter(posicao - 1)
            corrente = antecessor.proximo
            sucessor = corrente.proximo
            
            antecessor.proximo = sucessor        
        
        self.topo -= 1

    def imprimir(self):
        if self.topo > 0:
            corrente = self.cabeca

            print(corrente.dado, end=" ")

            while corrente.proximo != None:
                corrente = corrente.proximo
                print(corrente.dado, end=" ")

    def tamanho(self):
        return self.topo
    
    def inserirEmOrdem(self, valor):
        corrente = self.cabeca

        if self.topo == 0:
            self.inserir(valor, 0)
            return
        
        i = 0
        while i < self.topo:
            if valor < corrente.dado:
                break

            corrente = corrente.proximo
            i += 1

        self.inserir(valor, i)